src/main/kotlin/org/rust/lang/core/types/infer/Expressions.kt

```diff
@@ -132,16 +133,132 @@ fun inferExpressionType(expr: RsExpr): Ty {
+                dot2 != null && el.size == 0 -> "RangeFull" to null
+                dot2 != null && el.size == 1 -> {
+                    val e = el[0];
+                    if (e.startOffsetInParent < dot2.startOffsetInParent) {
+                        "RangeFrom" to e.type
+                    } else {
+                        "RangeTo" to e.type
+                    }
+                }
+                dot2 != null && el.size == 2 -> {
+                    "Range" to getMoreCompleteType(el[0].type, el[1].type)
+                }
+                dot3 != null && el.size == 1 -> {
+                    val e = el[0];
+                    if (e.startOffsetInParent < dot3.startOffsetInParent) {
+                        return TyUnknown
+                    } else {
+                        "RangeToInclusive" to e.type
+                    }
+                }
+                dot3 != null && el.size == 2 -> {
+                    "RangeInclusive" to getMoreCompleteType(el[0].type, el[1].type)
+                }
+
+                else -> return TyUnknown //Should never happen
+            }
+
+            val ty = tyFromAbsolutePath("std", "core", "ops::" + rangeName, expr);
+
+            if (rangeTy != null) {
+                val typeParameter = ty.getTypeParameter("Idx") ?: return ty
+                ty.substitute(mapOf(typeParameter to rangeTy))
+            } else {
+                ty
+            }
+        }
+
         is RsIndexExpr -> {
             val containerType = expr.containerExpr?.type ?: return TyUnknown
             val indexType = expr.indexExpr?.type ?: return TyUnknown
             findIndexOutputType(expr.project, containerType, indexType)
         }
 
+        is RsMacroExpr -> {
+            if (expr.vecMacro != null) {
+                val elements = expr.vecMacro!!.vecMacroArgs?.exprList ?: emptyList()
+                var elementType: Ty = TyUnknown;
+                for (e in elements) {
+                    elementType = getMoreCompleteType(e.type, elementType);
+                }
+
+                val ty = tyFromAbsolutePath("std", "collections", "vec::Vec", expr);
+                val tp = ty.getTypeParameter("T") ?: return ty
+
+                ty.substitute(mapOf(tp to elementType))
+            } else if (expr.logMacro != null) {
+                TyUnit
+            } else if (expr.macro != null) {
+                val macro = expr.macro ?: return TyUnknown;
+                when (macro) {
+                    is RsTryMacro -> {
+                        val base = macro.tryMacroArgs?.expr?.type ?: return TyUnknown;
+                        // This is super hackish. Need to figure out how to
+                        // identify known ty (See also the CString inspection).
+                        // Java uses fully qualified names for this, perhaps we
+                        // can do this as well? Will be harder to test though :(
+                        // See RsTryExpr where we handle the ? expression
```

Comment:
Yeah, the `?` was implemented before the `LangItems`. 

>I am not sure how to properly do it. Do you envision something like LangItems.isRange(type: Ty): Boolean or something like LangItems.isStdType(type: Ty, typeName: String): Boolean

Something like `fun isStdResult(ty: Ty)`. I don't think we need a super generic `typeName: String` variant, because there are only so many interesting types in Rust. The precise interface probably does not matter much at this stage: I think we can provide a nice API later, once we've collected all the use-cases for it in `LangItems.kt`.